### PR TITLE
Last Finalised State. MVP.

### LINKED ISSUE #3054: Node traverses the whole DAG on each replay
For some reason node traverses the whole DAG here on each replay https://github.com/rchain/rchain/blob/3e4ef77fdde277db4bb1397872283dec6f6af1ad/casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala#L421-L442, it takes about 20-30 sec on testnet (300k blocks in a chain).

We should get rid of it because its 
1. Slow
2. LFS assumes that we don't need and don't have that old blocks, so need to investigate why at all we do this.

### PR DESCRIPTION
**Pull request by nzpr**
Continuation of work on LFS.

Now we do not make a distinction between Approved Block from genesis ceremony and LFS. Approved Block is a block which state some set of validators is agree on. With the very first block set is defined by `required-sigs`. With all other blocks it is defined by safety oracle via finalization.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by nzpr** [comment]
With these changes I am able to connect to the network from LFS.

**Comment by nzpr** [comment]
This commit <https://github.com/rchain/rchain/pull/3055/commits/200272ad90e105d92b6485896d748bb72bc6581f> should be reverted to not mess with DagStore code and another way to prepare blocks for insertion should be implemented.

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala:152]: Please refactor this function using `tailRecM` to be stack safe.
https://typelevel.org/cats/typeclasses/monad.html#tailrecm

**Comment by nzpr** [comment]
Imo we don't need this validation at all 
https://github.com/rchain/rchain/blob/70cf529ca48f41126c235d535bfcf70894a34062/casper/src/main/scala/coop/rchain/casper/Casper.scala#L103-L128
- Ceremony master created block itself
- Genesis validators should check their signature on Approved block after they signed Unapproved block
- Usual peers should just rely on consensus on the state.
So we just need postStateHash of approved block

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala]: Because `Ordering[Long]` is defined there is more convenient function `sortBy` which only need selection of the field.
`blocksToAdd.flatten.sortBy(_.body.state.blockNumber)`

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala]: This is leftover from my previous changes, it's not used.

**Comment by tgrospic** [comment]
> Imo we don't need this validation at all
> 
> * Ceremony master created block itself
> * Usual peers should just rely on consensus on the state.
>   So we just need postStateHash of approved block

Yes! This is correct. Even for the very first genesis, state can be downloaded in the same way as for any (last) finalized block.

> * Genesis validators should check their signature on Approved block after they signed Unapproved block

Genesis validators still need to create the block to validate the hash and add their signature.

**Comment by nzpr** [comment]
> Genesis validators still need to create the block to validate the hash and add their signature.

Yes, but now genesis creates block (play+replay) - send to genesis validators - wait for approvals - and here again replay approved block. So we are removing the last step.
This also will make integration tests faster.

**Comment by nzpr** [comment]
Just want to push the naming once more :) Sorry for being pesky. 
Let's call it `approved block` - I believe `last finalised block` is a bad word, because each node might finalise blocks a bit differently, so we have to find block that is finalised on a majority of nodes. That is `approved Block`.

File: casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateBlockRequester.scala
Comment (zsluedem) [inline casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateBlockRequester.scala:127]: why 50? I remember we got a config for this. But I can not find it.

Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateBlockRequester.scala:127]: Its still hardcoded n Casper implementation, we are going to make it consistent <https://github.com/rchain/rchain/pull/2986>

File: casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala]: Typo in `heghts`.

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala]: Do you have some plans for this code or it can be removed?

**Comment by nzpr** [comment]
<https://github.com/rchain/rchain/pull/3055/checks?check_run_id=994259740#step:8:94> 
The only test is failing because dummy implementation of RSpace state manager is not complete. As we are creating tests in the follow up PR, we leave it for now.

**Review by tgrospic — APPROVED** [review]
LGTM

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala]: Yes, I think we should remove it

**Review by zsluedem — APPROVED** [review]
VEYR GOOD WORK! We will polish in more prs

File: casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala:132]: This will add to the tail blocks that are not in block store, eventually. Have to add some equivalent of `blockStored` from the previous version

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala:132]: All blocks should be in the blocks store downloaded in previous operation with `LastFinalizedStateBlockRequester`.

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala:132]: Oh, I see what you mean. There should be the check for last downloaded block where insertion should stop because we don't have all blocks to the first genesis.

Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala:132]: Fix is here #3092.

**Pull request merged**
